File: Catalog REST/src/main/java/cz/vhromada/catalog/rest/controller/GameController.java
package cz.vhromada.catalog.rest.controller;

import cz.vhromada.catalog.facade.GameFacade;
import cz.vhromada.catalog.facade.to.GameTO;
import cz.vhromada.validators.exceptions.RecordNotFoundException;
import cz.vhromada.validators.exceptions.ValidationException;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.ResponseBody;

/**
 * A class represents controller for games.
 *
 * @author Vladimir Hromada
 */
@Controller("gameController")
@RequestMapping("/games")
public class GameController extends JsonController {

	@Autowired
	@Qualifier("gameFacade")
	private GameFacade gameFacade;

	/** Creates new data. */
	@RequestMapping(value = "/new", method = RequestMethod.GET)
	@ResponseBody
	public void newData() {
		gameFacade.newData();
	}

	/**
	 * Returns list of games.
	 *
	 * @return list of games
	 */
	@RequestMapping(value = { "", "/", "list" }, method = RequestMethod.GET)
	@ResponseBody
	public String getGames() {
		return serialize(gameFacade.getGames());
	}

	/**
	 * Returns game with ID or null if there isn't such game.
	 *
	 * @param id ID
	 * @return game with ID or null if there isn't such game
	 * @throws IllegalArgumentException if ID is null
	 */
	@RequestMapping(value = "/{id}", method = RequestMethod.GET)
	@ResponseBody
	public String getGame(@PathVariable("id") final Integer id) {
		return serialize(gameFacade.getGame(id));
	}

	/**
	 * Adds game. Sets new ID and position.
	 *
	 * @param game game
	 * @throws IllegalArgumentException if game is null
	 * @throws ValidationException      if ID isn't null
	 *                                  or name is null
	 *                                  or name is empty string
	 *                                  or URL to english Wikipedia page about game is null
	 *                                  or URL to czech Wikipedia page about game is null
	 *                                  or count of media isn't positive number
	 *                                  or other data is null
	 *                                  or note is null
	 */
	@RequestMapping(value = "/add", method = RequestMethod.POST)
	@ResponseBody
	public void add(final String game) {
		gameFacade.add(deserialize(game, GameTO.class));
	}

	/**
	 * Updates game.
	 *
	 * @param game new value of game
	 * @throws IllegalArgumentException if game is null
	 * @throws ValidationException      if ID is null
	 *                                  or name is null
	 *                                  or name is empty string
	 *                                  or URL to english Wikipedia page about game is null
	 *                                  or URL to czech Wikipedia page about game is null
	 *                                  or count of media isn't positive number
	 *                                  or other data is null
	 *                                  or note is null
	 * @throws RecordNotFoundException  if game doesn't exist in data storage
	 */
	@RequestMapping(value = "/update", method = RequestMethod.POST)
	@ResponseBody
	public void update(final String game) {
		gameFacade.update(deserialize(game, GameTO.class));
	}

	/**
	 * Removes game.
	 *
	 * @param game game
	 * @throws IllegalArgumentException if game is null
	 * @throws ValidationException      if ID is null
	 * @throws RecordNotFoundException  if game doesn't exist in data storage
	 */
	@RequestMapping(value = "/remove", method = RequestMethod.POST)
	@ResponseBody
	public void remove(final String game) {
		gameFacade.remove(deserialize(game, GameTO.class));
	}

	/**
	 * Duplicates game.
	 *
	 * @param game game
	 * @throws IllegalArgumentException if game is null
	 * @throws ValidationException      if ID is null
	 * @throws RecordNotFoundException  if game doesn't exist in data storage
	 */
	@RequestMapping(value = "/duplicate", method = RequestMethod.POST)
	@ResponseBody
	public void duplicate(final String game) {
		gameFacade.duplicate(deserialize(game, GameTO.class));
	}

	/**
	 * Moves game in list one position up.
	 *
	 * @param game game
	 * @throws IllegalArgumentException if game is null
	 * @throws ValidationException      if ID is null
	 *                                  or game can't be moved up
	 * @throws RecordNotFoundException  if game doesn't exist in data storage
	 */
	@RequestMapping(value = "/moveUp", method = RequestMethod.POST)
	@ResponseBody
	public void moveUp(final String game) {
		gameFacade.moveUp(deserialize(game, GameTO.class));
	}

	/**
	 * Moves game in list one position down.
	 *
	 * @param game game
	 * @throws IllegalArgumentException if game is null
	 * @throws ValidationException      if ID is null
	 *                                  or game can't be moved down
	 * @throws RecordNotFoundException  if game doesn't exist in data storage
	 */
	@RequestMapping(value = "/moveDown", method = RequestMethod.POST)
	@ResponseBody
	public void moveDown(final String game) {
		gameFacade.moveDown(deserialize(game, GameTO.class));
	}

	/**
	 * Returns true if game exists.
	 *
	 * @param game game
	 * @return true if game exists
	 * @throws IllegalArgumentException if game is null
	 * @throws ValidationException      if ID is null
	 */
	@RequestMapping(value = "/exists", method = RequestMethod.POST)
	@ResponseBody
	public String exists(final String game) {
		return serialize(gameFacade.exists(deserialize(game, GameTO.class)));
	}

	/** Updates positions. */
	@RequestMapping(value = "/updatePositions", method = RequestMethod.GET)
	@ResponseBody
	public void updatePositions() {
		gameFacade.updatePositions();
	}

	/**
	 * Returns total count of media.
	 *
	 * @return total count of media
	 */
	@RequestMapping(value = "/totalMedia", method = RequestMethod.GET)
	@ResponseBody
	public String getTotalMediaCount() {
		return serialize(gameFacade.getTotalMediaCount());
	}

}

File: catalog-rest/src/main/kotlin/com/github/vhromada/catalog/rest/controller/GameController.kt
package com.github.vhromada.catalog.rest.controller

import com.github.vhromada.catalog.entity.Game
import com.github.vhromada.catalog.facade.GameFacade
import com.github.vhromada.common.web.controller.AbstractController
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestController

/**
 * A class represents controller for games.
 *
 * @author Vladimir Hromada
 */
@RestController("gameController")
@RequestMapping("/catalog/games")
class GameController(
    private val gameFacade: GameFacade
) : AbstractController() {

    /**
     * Returns list of games.
     *
     * @return list of games
     */
    @GetMapping
    fun getGames(): List<Game> {
        return processResult(result = gameFacade.getAll())!!
    }

    /**
     * Creates new data.
     */
    @PostMapping("/new")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun newData() {
        processResult(result = gameFacade.newData())
    }

    /**
     * Returns game with ID.
     *
     * @param id ID
     * @return game with ID
     */
    @GetMapping("/{id}")
    fun getGame(@PathVariable("id") id: Int): Game {
        return processResult(result = gameFacade.get(id = id))!!
    }

    /**
     * Adds game. Sets new ID and position.
     * <br></br>
     * Validation errors:
     *
     *  * ID isn't null
     *  * Position isn't null
     *  * Name is null
     *  * Name is empty string
     *  * URL to english Wikipedia page about game is null
     *  * URL to czech Wikipedia page about game is null
     *  * Count of media isn't positive number
     *  * Format is null
     *  * Other data is null
     *  * Note is null
     *
     * @param game game
     */
    @PutMapping("/add")
    @ResponseStatus(HttpStatus.CREATED)
    fun add(@RequestBody game: Game) {
        processResult(result = gameFacade.add(data = game))
    }

    /**
     * Updates game.
     * <br></br>
     * Validation errors:
     *
     *  * ID is null
     *  * Position is null
     *  * Name is null
     *  * Name is empty string
     *  * URL to english Wikipedia page about game is null
     *  * URL to czech Wikipedia page about game is null
     *  * Count of media isn't positive number
     *  * Format is null
     *  * Other data is null
     *  * Note is null
     *  * Game doesn't exist in data storage
     *
     * @param game new value of game
     */
    @PostMapping("/update")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun update(@RequestBody game: Game) {
        processResult(result = gameFacade.update(data = game))
    }

    /**
     * Removes game.
     * <br></br>
     * Validation errors:
     *
     *  * Game doesn't exist in data storage
     *
     * @param id ID
     */
    @DeleteMapping("/remove/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun remove(@PathVariable("id") id: Int) {
        processResult(result = gameFacade.remove(id = id))
    }

    /**
     * Duplicates game.
     * <br></br>
     * Validation errors:
     *
     *  * Game doesn't exist in data storage
     *
     * @param id ID
     */
    @PostMapping("/duplicate/{id}")
    @ResponseStatus(HttpStatus.CREATED)
    fun duplicate(@PathVariable("id") id: Int) {
        processResult(result = gameFacade.duplicate(id = id))
    }

    /**
     * Moves game in list one position up.
     * <br></br>
     * Validation errors:
     *
     *  * Game can't be moved up
     *  * Game doesn't exist in data storage
     *
     * @param id ID
     */
    @PostMapping("/moveUp/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun moveUp(@PathVariable("id") id: Int) {
        processResult(result = gameFacade.moveUp(id = id))
    }

    /**
     * Moves game in list one position down.
     * <br></br>
     * Validation errors:
     *
     *  * Game can't be moved down
     *  * Game doesn't exist in data storage
     *
     * @param id ID
     */
    @PostMapping("/moveDown/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun moveDown(@PathVariable("id") id: Int) {
        processResult(result = gameFacade.moveDown(id = id))
    }

    /**
     * Updates positions.
     */
    @PostMapping("/updatePositions")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun updatePositions() {
        processResult(result = gameFacade.updatePositions())
    }

    /**
     * Returns total count of media.
     *
     * @return total count of media
     */
    @GetMapping("/totalMedia")
    fun totalMediaCount(): Int {
        return processResult(result = gameFacade.getTotalMediaCount())!!
    }

}
